### ISSUE TITLE
Socket timeouts and heartbeats

### ISSUE THREAD

**Issue by IGx89** [kushagharahi/homebridge-dahua-alerts#14]
Hello! After enabling debug I noticed a lot of lines like the following in my logs:

```
[30/03/2021, 20:14:33] [dahua-alerts] Socket connection timed out or closed by NVR.
[30/03/2021, 20:14:33] [dahua-alerts] Reconnecting in 1s.
[30/03/2021, 20:15:04] [dahua-alerts] Socket connection timed out or closed by NVR.
[30/03/2021, 20:15:04] [dahua-alerts] Reconnecting in 1s.
[30/03/2021, 20:15:35] [dahua-alerts] Socket connection timed out or closed by NVR.
[30/03/2021, 20:15:35] [dahua-alerts] Reconnecting in 1s.
```

Best I can tell from looking at your code, your approach is to listen for up to 30 seconds for an event and if none arrives, you time out the socket and then reconnect a second later. Is that intentional?

Based on the API docs, you have two other options: sending a keepalive string every 0-60 seconds; or telling the device to send a heartbeat every 0-60 seconds. The docs recommend the latter, which I successfully tested by adding `&heartbeat=15` to the API call. When I do that the following happens instead:

```
[30/03/2021, 19:51:21] [dahua-alerts] Response recieved: 
--myboundary
Content-Type: text/plain
Content-Length:9

Heartbeat

[30/03/2021, 19:51:36] [dahua-alerts] Response recieved: 
--myboundary
Content-Type: text/plain
Content-Length:9

Heartbeat

[30/03/2021, 19:51:51] [dahua-alerts] Response recieved: 
--myboundary
Content-Type: text/plain
Content-Length:9

Heartbeat

[30/03/2021, 19:52:06] [dahua-alerts] Response recieved: 
--myboundary
Content-Type: text/plain
Content-Length:9

Heartbeat
```

That seems to be much better. Your thoughts? This is on a new-ish N43AJ52 so there's a chance it doesn't work on older devices, though it sounds like you have an older device so you could confirm if it does or not if you wanted.

Nothing here has any functional impact that I'm aware of, though the current way it works does indicate that 2 seconds out of every minute it would drop events due to the 1 second reconnect delay, leading to a 3.33% failure rate.

**Comment by kushagharahi** [comment]
Thanks for this detailed report! 

It looks like adding `&heartbeat=15` doesn't do anything with my NVR however it doesn't cause any issues either. So it could be something that we can add without (potentially) breaking cameras/NVRs that don't support this option.

BUT

>Based on the API docs, you have two other options: sending a keepalive string every 0-60 seconds

[This plugin theoretically sends a KeepAlive string every second](https://github.com/kushagharahi/homebridge-dahua-alerts/blob/master/src/dahua.ts#L20) with Node's http/https agent

So we need to find out if Node's http/https agent is actually sending the KeepAlive string and why it doesn't work. If that investigation isn't fruitful, then perhaps adding `&heartbeat=15` is the solution. I'd prefer the KeepAlive route because that should theoretically work across all models.

**Comment by IGx89** [comment]
I previously did some digging to see what could be done to make keepalive work and ran into a brick wall. Node's implementation seems designed for high concurrency, just short-term connection pooling. According to [the docs](https://nodejs.org/api/net.html#net_socket_setkeepalive_enable_initialdelay), your settings theoretically will cause it to send a keepalive packet after 1 second, then 9 more (until TCP_KEEPCNT is hit) at which point it stops.

I haven't confirmed that behavior using a packet sniffer, but it would explain what I'm seeing. I stopped digging further into it because I discovered that changing TCP_KEEPCNT/TCP_KEEPINTVL isn't natively supported in Node and and if you use [a native modul](https://www.npmjs.com/package/net-keepalive)e to change them (like [node-dahua-api](https://github.com/nayrnet/node-dahua-api) does) it breaks Windows support (my development OS).

Since that ended up being a dead end, I've been successfully using the much less complicated heartbeat route (20 seconds). Just adding that to the URL wasn't quite good enough though I discovered. It would occasionally (one or twice a day) still drop the connection and take minutes to reconnect. I solved that issue by changing a couple http agent settings: maxFreeSockets from 0 to 1, and timeout from 30000 to 60000. Not sure which one did the trick, but it's been working without fail for weeks so far so I haven't touched it further.

**Comment by kushagharahi** [comment]
>I discovered that changing TCP_KEEPCNT/TCP_KEEPINTVL isn't natively supported in Node and and if you use a native module to change them (like node-dahua-api does) it breaks Windows support (my development OS)

That stinks. I'm going to let this stew as an open issue to see if a good alternative comes up since the `heartbeat` url param seems to be a newer feature.

The `heartbeat` param addition doesn't break my NVR but I'm afraid other systems that don't have it enabled might have their plugin break due to the unknown param.

**Comment by magic7s** [comment]
I'm getting a lot of errors (but motion detection is working). How can I test the `&heartbeat=15`? I don't know how to load a branch on my homebridge server. 

edit: I found a quick way to test it and still getting the scrolling errors. I did a pcap and the plugin is trying to reach out to the camera every 30 seconds and getting 40X error. My guess is it is trying to reuse a key from another session, but that is a guess. I'm just going to ignore it for now as it's functionally working.

**Comment by kushagharahi** [comment]
This plugin works on a socket based connection to the camera. Please open a new issue instead of commenting on unrelated existing issues next time @magic7s If you are getting errors and then reconnection issues, I highly doubt the plugin is functioning properly.

**Comment by kushagharahi** [comment]
@magic7s found a way of removing the socket timeouts altogether to rely only on the KeepAlive packets. Thanks to their discovery this issue is resolved in 1.3.0
